geo/src/algorithm/voronoi_diagram.rs

```diff
@@ -0,0 +1,970 @@
+    DelaunayTriangle, DelaunayTriangulationError, DEFAULT_SUPER_TRIANGLE_EXPANSION,
+};
+use crate::{BoundingRect, TriangulateDelaunay};
+
+type Result<T> = result::Result<T, VoronoiDiagramError>;
+
+#[derive(Debug, Clone)]
+pub struct VoronoiComponents<T: GeoFloat> {
+    pub delaunay_triangles: Vec<Triangle<T>>,
+    pub vertices: Vec<Coord<T>>,
+    pub lines: Vec<Line<T>>,
+    pub neighbours: Vec<(Option<usize>, Option<usize>)>,
+}
+
+pub trait VoronoiDiagram<T: GeoFloat>
+where
+    f64: From<T>,
+{
+    fn compute_voronoi_components(
+        &self,
+        clipping_mask: Option<&Polygon<T>>,
+    ) -> Result<VoronoiComponents<T>>;
+    fn compute_voronoi_diagram(&self, clipping_mask: Option<&Polygon<T>>) -> Result<Vec<Polygon>>;
+}
+
+impl<T: GeoFloat> VoronoiDiagram<T> for Polygon<T>
+where
+    f64: From<T>,
+{
+    fn compute_voronoi_components(
+        &self,
+        clipping_mask: Option<&Polygon<T>>,
+    ) -> Result<VoronoiComponents<T>> {
+        compute_voronoi_components(self, clipping_mask)
+    }
+
+    fn compute_voronoi_diagram(&self, _clipping_mask: Option<&Polygon<T>>) -> Result<Vec<Polygon>> {
+        todo!("Need to map the components to voronoi cells");
+    }
+}
+impl<T: GeoFloat> VoronoiDiagram<T> for MultiPoint<T>
+where
+    f64: From<T>,
+{
+    fn compute_voronoi_components(
+        &self,
+        clipping_mask: Option<&Polygon<T>>,
+    ) -> Result<VoronoiComponents<T>> {
+        compute_voronoi_components(self, clipping_mask)
+    }
+
+    fn compute_voronoi_diagram(&self, _clipping_mask: Option<&Polygon<T>>) -> Result<Vec<Polygon>> {
+        todo!("Need to map the components to voronoi cells");
+    }
+}
+
+fn compute_voronoi_components<T: GeoFloat, U: TriangulateDelaunay<T>>(
+    polygon: &U,
+    clipping_mask: Option<&Polygon<T>>,
+) -> Result<VoronoiComponents<T>>
+where
+    f64: From<T>,
+{
+    let triangles = polygon
+        .delaunay_triangulation()
+        .map_err(VoronoiDiagramError::DelaunayError)?;
+
+    compute_voronoi_components_from_delaunay(&triangles, clipping_mask)
+}
+
+/// Compute the Voronoi Diagram from Delaunay Triangles
+/// The Voronoi Diagram is a [dual graph](https://en.wikipedia.org/wiki/Dual_graph)
+/// of the [Delaunay Triangulation](https://en.wikipedia.org/wiki/Delaunay_triangulation)
+/// and thus the Voronoi Diagram can be created from the Delaunay Triangulation.
+pub fn compute_voronoi_components_from_delaunay<T: GeoFloat>(
+    triangles: &[Triangle<T>],
+    clipping_mask: Option<&Polygon<T>>,
+) -> Result<VoronoiComponents<T>>
+where
+    f64: From<T>,
+{
+    let delaunay_triangles: Vec<DelaunayTriangle<T>> =
+        triangles.iter().map(|x| (*x).into()).collect();
+
+    if delaunay_triangles.is_empty() {
+        return Ok(VoronoiComponents {
+            delaunay_triangles: vec![],
+            vertices: vec![],
+            lines: vec![],
+            neighbours: vec![],
+        });
+    }
+
+    let clipping_mask = match clipping_mask {
```

Comment:
IMO this should be moved to be the first statement in the fn, as this `clipping_mask` var shadows the argument of the same name.  I dont mind shadowing, if it is done at the beginning, so then the reader can forget about the arg of the same name.